### ISSUE TITLE
ALTER KEYSPACE 'keyspace_new_dc' Fails with "Keyspace Doesn't Exist" Error

### ISSUE THREAD

**Issue by timtimb0t** [scylladb/scylladb#21337]
## Packages

Scylla version: `6.3.0~dev-20241018.b11d50f59191` with build-id `d5fe38f8fd12d9b834688320151f36fb8c1e050d`

Kernel Version: `6.8.0-1017-aws`

## Issue description

New issue

Test creates new node in new dc with new keyspace and tries to apply the replication strategy for all nodes (including new) and run repair for all nodes. Afterwards, during the altering process (which was initiated 14:39:54)  there is an error occurred:

```
2024-10-19T14:39:54.873+00:00 longevity-50gb-12h-master-db-node-b9918db7-2      !ERR | scylla[5487]:  [shard 0:stmt] query_processor - CQL statement "ALTER KEYSPACE keyspace_new_dc WITH replication = {'class': 'NetworkTopologyStrategy', 'eu-west': 3}" with topology request_id "6548590c-8e27-11ef-a0fc-77ce44221497" failed with error: "Can't ALTER keyspace keyspace_new_dc, keyspace doesn't exist"
```

also, there is interesting log entry right before the error (the test should not have started decommissioning by this time according to code):

```
2024-10-19T14:39:54.872+00:00 longevity-50gb-12h-master-db-node-b9918db7-2     !INFO | scylla[5487]:  [shard 0: gms] gossip - Removed endpoint 10.4.8.86
2024-10-19T14:39:54.872+00:00 longevity-50gb-12h-master-db-node-b9918db7-2     !INFO | scylla[5487]:  [shard 0: gms] gossip - InetAddress f5d0ac9c-5d48-4e26-9480-9332c4cf7307/10.4.8.86 is now DOWN, status = NORMAL
```

New node decommission was initiated at 14:36:46:
```
2024-10-19T14:36:46.050+00:00 longevity-50gb-12h-master-db-node-b9918db7-5     !INFO | scylla[5711]:  [shard  1:strm] api - decommission
```

## Impact

No visible impact, scylla is alive

## How frequently does it reproduce?
few times by this moment

*Describe the frequency with how this issue can be reproduced.*

## Installation details

Cluster size: 4 nodes (i4i.4xlarge)

Scylla Nodes used in this run:
- longevity-50gb-12h-master-db-node-b9918db7-7 (3.250.38.94 | 10.4.10.238) (shards: 12)
- longevity-50gb-12h-master-db-node-b9918db7-6 (52.209.198.119 | 10.4.11.99) (shards: 11)
- longevity-50gb-12h-master-db-node-b9918db7-5 (34.244.99.121 | 10.4.8.86) (shards: 13)
- longevity-50gb-12h-master-db-node-b9918db7-4 (52.212.34.111 | 10.4.8.104) (shards: 14)
- longevity-50gb-12h-master-db-node-b9918db7-3 (34.242.246.82 | 10.4.10.219) (shards: 13)
- longevity-50gb-12h-master-db-node-b9918db7-2 (3.252.87.187 | 10.4.9.52) (shards: 10)
- longevity-50gb-12h-master-db-node-b9918db7-1 (3.249.197.56 | 10.4.8.194) (shards: 10)

OS / Image: `ami-0d68a690456025633` (aws: undefined_region)

Test: `longevity-150gb-asymmetric-cluster-12h-test`
Test id: `b9918db7-e911-46bc-b4ea-9966ab48077f`
Test name: `scylla-master/tier1/longevity-150gb-asymmetric-cluster-12h-test`
Test method: `longevity_test.LongevityTest.test_custom_time`
Test config file(s):

- [longevity-150GB-12h-autorization-LimitedMonkey.yaml](https://github.com/scylladb/scylla-cluster-tests/blob/99d539bbdc3a4ec912511e761275f0c202fd9a30/test-cases/longevity/longevity-150GB-12h-autorization-LimitedMonkey.yaml)

<details>
<summary>
Logs and commands
</summary>

- Restore Monitor Stack command: `$ hydra investigate show-monitor b9918db7-e911-46bc-b4ea-9966ab48077f`
- Restore monitor on AWS instance using [Jenkins job](https://jenkins.scylladb.com/view/QA/job/QA-tools/job/hydra-show-monitor/parambuild/?test_id=b9918db7-e911-46bc-b4ea-9966ab48077f)
- Show all stored logs command: `$ hydra investigate show-logs b9918db7-e911-46bc-b4ea-9966ab48077f`

## Logs:
- **longevity-50gb-12h-master-db-node-b9918db7-5** - [https://cloudius-jenkins-test.s3.amazonaws.com/b9918db7-e911-46bc-b4ea-9966ab48077f/20241019_031103/longevity-50gb-12h-master-db-node-b9918db7-5-b9918db7.tar.gz](https://cloudius-jenkins-test.s3.amazonaws.com/b9918db7-e911-46bc-b4ea-9966ab48077f/20241019_031103/longevity-50gb-12h-master-db-node-b9918db7-5-b9918db7.tar.gz)
- **longevity-50gb-12h-master-db-node-b9918db7-6** - [https://cloudius-jenkins-test.s3.amazonaws.com/b9918db7-e911-46bc-b4ea-9966ab48077f/20241019_031103/longevity-50gb-12h-master-db-node-b9918db7-6-b9918db7.tar.gz](https://cloudius-jenkins-test.s3.amazonaws.com/b9918db7-e911-46bc-b4ea-9966ab48077f/20241019_031103/longevity-50gb-12h-master-db-node-b9918db7-6-b9918db7.tar.gz)
- **longevity-50gb-12h-master-db-node-b9918db7-1** - [https://cloudius-jenkins-test.s3.amazonaws.com/b9918db7-e911-46bc-b4ea-9966ab48077f/20241019_031103/longevity-50gb-12h-master-db-node-b9918db7-1-b9918db7.tar.gz](https://cloudius-jenkins-test.s3.amazonaws.com/b9918db7-e911-46bc-b4ea-9966ab48077f/20241019_031103/longevity-50gb-12h-master-db-node-b9918db7-1-b9918db7.tar.gz)
- **db-cluster-b9918db7.tar.gz** - [https://cloudius-jenkins-test.s3.amazonaws.com/b9918db7-e911-46bc-b4ea-9966ab48077f/20241020_033346/db-cluster-b9918db7.tar.gz](https://cloudius-jenkins-test.s3.amazonaws.com/b9918db7-e911-46bc-b4ea-9966ab48077f/20241020_033346/db-cluster-b9918db7.tar.gz)
- **sct-runner-events-b9918db7.tar.gz** - [https://cloudius-jenkins-test.s3.amazonaws.com/b9918db7-e911-46bc-b4ea-9966ab48077f/20241020_033346/sct-runner-events-b9918db7.tar.gz](https://cloudius-jenkins-test.s3.amazonaws.com/b9918db7-e911-46bc-b4ea-9966ab48077f/20241020_033346/sct-runner-events-b9918db7.tar.gz)
- **sct-b9918db7.log.tar.gz** - [https://cloudius-jenkins-test.s3.amazonaws.com/b9918db7-e911-46bc-b4ea-9966ab48077f/20241020_033346/sct-b9918db7.log.tar.gz](https://cloudius-jenkins-test.s3.amazonaws.com/b9918db7-e911-46bc-b4ea-9966ab48077f/20241020_033346/sct-b9918db7.log.tar.gz)
- **loader-set-b9918db7.tar.gz** - [https://cloudius-jenkins-test.s3.amazonaws.com/b9918db7-e911-46bc-b4ea-9966ab48077f/20241020_033346/loader-set-b9918db7.tar.gz](https://cloudius-jenkins-test.s3.amazonaws.com/b9918db7-e911-46bc-b4ea-9966ab48077f/20241020_033346/loader-set-b9918db7.tar.gz)
- **monitor-set-b9918db7.tar.gz** - [https://cloudius-jenkins-test.s3.amazonaws.com/b9918db7-e911-46bc-b4ea-9966ab48077f/20241020_033346/monitor-set-b9918db7.tar.gz](https://cloudius-jenkins-test.s3.amazonaws.com/b9918db7-e911-46bc-b4ea-9966ab48077f/20241020_033346/monitor-set-b9918db7.tar.gz)

[Jenkins job URL](https://jenkins.scylladb.com/job/scylla-master/job/tier1/job/longevity-150gb-asymmetric-cluster-12h-test/41/)
[Argus](https://argus.scylladb.com/test/7f604213-5cad-4418-921e-1ca98918071c/runs?additionalRuns[]=b9918db7-e911-46bc-b4ea-9966ab48077f)
</details>

**Comment by timtimb0t** [comment]
Need more investigation, will update this issue soon

**Comment by timtimb0t** [comment]
@pehala , could you please help with investigation of this issue? I'm not sure if root cause of such behavior is async python code

**Comment by pehala** [comment]
@timtimb0t do we know why the decomission happened? Was it due to a nemesis?

**Comment by yarongilor** [comment]
@timtimb0t , please consider elaborating about the relevant steps of the add-remove-dc nemesis.

i see this nemesis ran twice, failing on the first time and passing on the second one.
the second error seen (in Argus) may point to a problem:
```
cassandra.OperationTimedOut: errors={<Host: 10.4.10.219:9042 eu-west>: ConnectionException('Host has been marked down or removed'), <Host: 10.4.8.104:9042 eu-west>: ConnectionException('Host has been marked down or removed'), <Host: 10.4.8.194:9042 eu-west>: ConnectionException('Host has been marked down or removed')}, last_host=10.4.9.52:9042
```
the other nodes refused to talk with node-2 about this alter-keyspace, for some while, or it just takes more time than this nemesis waits before dropping the keyspace. (or it's both issues, not sure).
so it looks like the keyspace was already dropped before alter ended:
```
< t:2024-10-19 14:35:32,558 f:common.py       l:1324 c:utils                p:DEBUG > Executing CQL 'ALTER KEYSPACE keyspace_new_dc WITH replication = {'class': 'NetworkTopologyStrategy', 'eu-west': 3}' ...
< t:2024-10-19 14:36:34,037 f:common.py       l:1316 c:utils                p:DEBUG > Executing CQL 'DROP KEYSPACE IF EXISTS keyspace_new_dc' ...

```
```
< t:2024-10-19 14:36:39,219 f:thread.py       l:58   c:cassandra.cluster    p:DEBUG > Refreshing schema in response to schema change. {'target_type': 'KEYSPACE', 'change_type': 'DROPPED', 'keyspace': 'keyspace_new_dc'} )
```

@soyacz , does `temporary_replication_strategy_setter` wait or verify successful altering of keyspace before drop? we may consider adding it if not.

@kbr-scylla , if relevant, i wonder if this alter-keyspace failure could be related to https://github.com/scylladb/scylladb/issues/20282#issuecomment-2444021905.

p.s. i'm also not sure why `"eu-west_nemesis_dc": "1"` in this message: 
```
< t:2024-10-19 21:14:04,282 f:db_log_reader.py l:125  c:sdcm.db_log_reader   p:DEBUG > 2024-10-19T21:14:04.203+00:00 longevity-50gb-12h-master-db-node-b9918db7-3     !INFO | scylla[974]:  [shard  0: gms] migration_manager - Update Keyspace: KSMetaData{name=keyspace_new_dc, strategyClass=org.apache.cassandra.locator.NetworkTopologyStrategy, strategyOptions={"eu-west": "3", "eu-west_nemesis_dc": "1"}, cfMetaData={}, durable_writes=true, tablets={"enabled":false}, userTypes=org.apache.cassandra.config.UTMetaData@0x600004d185e0}
```

**Comment by mykaul** [comment]
I see `tablets={"enabled":false}` - but I want to make sure - this is WITHOUT tablets, right?

**Comment by timtimb0t** [comment]
reproduced there:
## Packages

Scylla version: `6.3.0~dev-20241115.5bc03da0c4c1` with build-id `616f46f4e7794bef8367877dfce801c0ddae5371`





Kernel Version: `6.8.0-1018-aws`


## Installation details

Cluster size: 4 nodes (i4i.4xlarge)

Scylla Nodes used in this run:
- longevity-50gb-12h-master-db-node-a02314fa-7 (34.245.192.122 | 10.4.9.150) (shards: 10)
- longevity-50gb-12h-master-db-node-a02314fa-6 (34.253.212.236 | 10.4.11.151) (shards: 11)
- longevity-50gb-12h-master-db-node-a02314fa-5 (3.255.153.21 | 10.4.8.214) (shards: 14)
- longevity-50gb-12h-master-db-node-a02314fa-4 (54.170.228.137 | 10.4.10.229) (shards: 12)
- longevity-50gb-12h-master-db-node-a02314fa-3 (3.254.53.79 | 10.4.10.35) (shards: 8)
- longevity-50gb-12h-master-db-node-a02314fa-2 (3.250.154.119 | 10.4.8.110) (shards: 13)
- longevity-50gb-12h-master-db-node-a02314fa-1 (54.154.51.96 | 10.4.10.107) (shards: 11)


OS / Image: `ami-01b1698557082262f` (aws: undefined_region)

Test: `longevity-150gb-asymmetric-cluster-12h-test`
Test id: `a02314fa-5fa4-4994-af23-20e5bb2f8183`
Test name: `scylla-master/tier1/longevity-150gb-asymmetric-cluster-12h-test`
Test method: `longevity_test.LongevityTest.test_custom_time`
Test config file(s):

- [longevity-150GB-12h-autorization-LimitedMonkey.yaml](https://github.com/scylladb/scylla-cluster-tests/blob/3c880d356e617d2de7b7ecaa53ab4d00c7061fd1/test-cases/longevity/longevity-150GB-12h-autorization-LimitedMonkey.yaml)


<details>
<summary>
Logs and commands
</summary>


- Restore Monitor Stack command: `$ hydra investigate show-monitor a02314fa-5fa4-4994-af23-20e5bb2f8183`
- Restore monitor on AWS instance using [Jenkins job](https://jenkins.scylladb.com/view/QA/job/QA-tools/job/hydra-show-monitor/parambuild/?test_id=a02314fa-5fa4-4994-af23-20e5bb2f8183)
- Show all stored logs command: `$ hydra investigate show-logs a02314fa-5fa4-4994-af23-20e5bb2f8183`


## Logs:
- **longevity-50gb-12h-master-db-node-a02314fa-5** - [https://cloudius-jenkins-test.s3.amazonaws.com/a02314fa-5fa4-4994-af23-20e5bb2f8183/20241116_041041/longevity-50gb-12h-master-db-node-a02314fa-5-a02314fa.tar.gz](https://cloudius-jenkins-test.s3.amazonaws.com/a02314fa-5fa4-4994-af23-20e5bb2f8183/20241116_041041/longevity-50gb-12h-master-db-node-a02314fa-5-a02314fa.tar.gz)
- **longevity-50gb-12h-master-db-node-a02314fa-6** - [https://cloudius-jenkins-test.s3.amazonaws.com/a02314fa-5fa4-4994-af23-20e5bb2f8183/20241116_041041/longevity-50gb-12h-master-db-node-a02314fa-6-a02314fa.tar.gz](https://cloudius-jenkins-test.s3.amazonaws.com/a02314fa-5fa4-4994-af23-20e5bb2f8183/20241116_041041/longevity-50gb-12h-master-db-node-a02314fa-6-a02314fa.tar.gz)
- **longevity-50gb-12h-master-db-node-a02314fa-3** - [https://cloudius-jenkins-test.s3.amazonaws.com/a02314fa-5fa4-4994-af23-20e5bb2f8183/20241116_041041/longevity-50gb-12h-master-db-node-a02314fa-3-a02314fa.tar.gz](https://cloudius-jenkins-test.s3.amazonaws.com/a02314fa-5fa4-4994-af23-20e5bb2f8183/20241116_041041/longevity-50gb-12h-master-db-node-a02314fa-3-a02314fa.tar.gz)
- **core.scylla-longevity-50gb-12h-master-db-node-a02314fa-1-2024-11-17_01-59-18.gz** - [https://storage.cloud.google.com/upload.scylladb.com/core.scylla.106.057eb3b1ea86430da20325f100777a7e.935.1731805840000000./core.scylla.106.057eb3b1ea86430da20325f100777a7e.935.1731805840000000.zst](https://storage.cloud.google.com/upload.scylladb.com/core.scylla.106.057eb3b1ea86430da20325f100777a7e.935.1731805840000000./core.scylla.106.057eb3b1ea86430da20325f100777a7e.935.1731805840000000.zst)
- **db-cluster-a02314fa.tar.gz** - [https://cloudius-jenkins-test.s3.amazonaws.com/a02314fa-5fa4-4994-af23-20e5bb2f8183/20241117_043151/db-cluster-a02314fa.tar.gz](https://cloudius-jenkins-test.s3.amazonaws.com/a02314fa-5fa4-4994-af23-20e5bb2f8183/20241117_043151/db-cluster-a02314fa.tar.gz)
- **sct-runner-events-a02314fa.tar.gz** - [https://cloudius-jenkins-test.s3.amazonaws.com/a02314fa-5fa4-4994-af23-20e5bb2f8183/20241117_043151/sct-runner-events-a02314fa.tar.gz](https://cloudius-jenkins-test.s3.amazonaws.com/a02314fa-5fa4-4994-af23-20e5bb2f8183/20241117_043151/sct-runner-events-a02314fa.tar.gz)
- **sct-a02314fa.log.tar.gz** - [https://cloudius-jenkins-test.s3.amazonaws.com/a02314fa-5fa4-4994-af23-20e5bb2f8183/20241117_043151/sct-a02314fa.log.tar.gz](https://cloudius-jenkins-test.s3.amazonaws.com/a02314fa-5fa4-4994-af23-20e5bb2f8183/20241117_043151/sct-a02314fa.log.tar.gz)
- **loader-set-a02314fa.tar.gz** - [https://cloudius-jenkins-test.s3.amazonaws.com/a02314fa-5fa4-4994-af23-20e5bb2f8183/20241117_043151/loader-set-a02314fa.tar.gz](https://cloudius-jenkins-test.s3.amazonaws.com/a02314fa-5fa4-4994-af23-20e5bb2f8183/20241117_043151/loader-set-a02314fa.tar.gz)
- **monitor-set-a02314fa.tar.gz** - [https://cloudius-jenkins-test.s3.amazonaws.com/a02314fa-5fa4-4994-af23-20e5bb2f8183/20241117_043151/monitor-set-a02314fa.tar.gz](https://cloudius-jenkins-test.s3.amazonaws.com/a02314fa-5fa4-4994-af23-20e5bb2f8183/20241117_043151/monitor-set-a02314fa.tar.gz)


[Jenkins job URL](https://jenkins.scylladb.com/job/scylla-master/job/tier1/job/longevity-150gb-asymmetric-cluster-12h-test/46/)
[Argus](https://argus.scylladb.com/test/7f604213-5cad-4418-921e-1ca98918071c/runs?additionalRuns[]=a02314fa-5fa4-4994-af23-20e5bb2f8183)
</details>

**Comment by mykaul** [comment]
@timtimb0t - any additional information in the logs that can be helpful here?

**Comment by kbr-scylla** [comment]
Apparently the ALTER was requested before the DROP, hence the confusion.

ALTER starts at:
```
< t:2024-10-19 14:35:32,558 f:common.py       l:1324 c:utils                p:DEBUG > Executing CQL 'ALTER KEYSPACE keyspace_new_dc WITH replication = {'class': 'NetworkTopologyStrategy', 'eu-west': 3}' ...
```

DROP starts at:
```
< t:2024-10-19 14:36:34,037 f:common.py       l:1324 c:utils                p:DEBUG > Executing CQL 'DROP KEYSPACE IF EXISTS keyspace_new_dc' ...
```

-- about a minute later.


However, the two operations are happening **concurrently**. If you have concurrent operations, there is no guarantee in which order they will execute -- the time of the request doesn't matter, because the operation that was requested earlier might actually arrive and get executed later.

If you want to guarantee order of execution, you must wait until one of the operation finishes before starting the next one.

So this is a test problem.

---

A different question here is why it took 4 minutes for the ALTER to actually get executed? It looks that the DROP arrived at node-2 (because it printed the error), perhaps node-2 was busy and the operation was waiting on a lock before it could actually start. Or node-2 was somehow partitioned from the cluster or had connectivity problems during this time.

@yarongilor observed:
> the other nodes refused to talk with node-2 about this alter-keyspace, for some while, or it just takes more time than this nemesis waits before dropping the keyspace. (or it's both issues, not sure).

apparently node-2 has general problems during this time, not being able to talk to the rest of the cluster. Maybe it was overloaded. It was very busy with compactions during this time.

---

Anyway, I'm closing the issue, the test or nemesis must be adjusted -- if it expects operations to happen in certain order, it must serialize them so. @timtimb0t

**Comment by kbr-scylla** [comment]
> @soyacz , does temporary_replication_strategy_setter wait or verify successful altering of keyspace before drop? we may consider adding it if not.

^^^

**Comment by soyacz** [comment]
> > [@soyacz](https://github.com/soyacz) , does temporary_replication_strategy_setter wait or verify successful altering of keyspace before drop? we may consider adding it if not.
> 
> ^^^

Sorry, I didn't spot this question.
No, it's not waiting and I see issue was already raised: https://github.com/scylladb/scylla-cluster-tests/issues/9430
